scipio/src/channels/local_channel.rs

```diff
@@ -0,0 +1,586 @@
+    /// ex.run(async move {
+    ///     let (sender, mut receiver) = LocalChannel::new_bounded(1);
+    ///     sender.send(0);
+    ///     sender.send(0).unwrap_err(); // no more capacity
+    ///     receiver.next().await.unwrap(); // now we have capacity again
+    ///     drop(receiver); // but because the receiver is destroyed send will err
+    ///     sender.send(0).unwrap_err();
+    /// });
+    /// ```
+    ///
+    /// [`BrokenPipe`]: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.BrokenPipe
+    /// [`Other`]: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.Other
+    /// [`io::Error`]: https://doc.rust-lang.org/std/io/struct.Error.html
+    pub fn send(&self, item: T) -> io::Result<()> {
+        self.channel.push(item)
+    }
+
+    /// Sends data into this channel when it is ready to receive it
+    ///
+    /// For an unbounded channel this is just a more expensive version of [`send`]. Prefer
+    /// to use [`send`] instead.
+    ///
+    /// For a bounded channel this will push to the channel when the channel is ready to
+    /// receive data.
+    ///
+    /// # Examples
+    /// ```
+    /// use scipio::{LocalExecutor, Local};
+    /// use scipio::channels::LocalChannel;
+    ///
+    /// let ex = LocalExecutor::make_default();
+    /// ex.run(async move {
+    ///     let (sender, receiver) = LocalChannel::new_bounded(1);
+    ///     sender.send_eventually(0).await.unwrap();
+    ///     drop(receiver);
+    /// });
+    /// ```
+    ///
+    /// [`send`]: struct.LocalSender.html#method.send
+    pub async fn send_eventually(&self, item: T) -> io::Result<()> {
+        if self.has_room() {
+            self.send(item)
+        } else {
+            let waiter = SendWaiter {
+                channel: LocalChannel {
+                    state: self.channel.state.clone(),
+                },
+            };
+            waiter.await;
+            self.send(item)
+        }
+    }
+
+    /// Checks if there is room to send data in this channel
+    ///
+    /// # Examples
+    /// ```
+    /// use scipio::{LocalExecutor, Local};
+    /// use scipio::channels::LocalChannel;
+    ///
+    /// let ex = LocalExecutor::make_default();
+    /// ex.run(async move {
+    ///     let (sender, receiver) = LocalChannel::new_bounded(1);
+    ///     assert_eq!(sender.has_room(), true);
+    ///     sender.send(0);
+    ///     assert_eq!(sender.has_room(), false);
+    ///     drop(receiver);
+    /// });
+    /// ```
+    pub fn has_room(&self) -> bool {
+        self.channel.has_room()
+    }
+}
+
+impl<T> Drop for LocalSender<T> {
+    fn drop(&mut self) {
+        let mut state = self.channel.state.borrow_mut();
+        state.sender_alive = false;
+
+        if !state.recv_waiters.is_empty() {
+            let waiters = std::mem::replace(&mut state.recv_waiters, Vec::new());
+            drop(state);
+
+            for w in waiters {
+                w.wake();
+            }
+        }
+    }
+}
```

Comment:
Can a `LocalSender` be dropped while there are still `send_waiters`? Something like calling `send_eventually` without awaiting the result.

Reply:
I spent some time thinking about this.

I don't think it should be possible:
- Rust futures are lazy, so there is no such thing as "not waiting on them". If you don't .await they don't execute.
- For that situation to trigger you have to spawn a new detached task and have the `LocalSender` and the send operation running in different tasks
- But for you to pass that into a spawn-like method you need an `Rc` pointer: spawn-like methods require static lifetime for all references passed to it.
- Which means if you drop at this point you'd be dropping the reference count. The senders would keep the refcount up and prevent an early drop.